src/Text/Gigaparsec/Internal/Token/Space.hs

```diff
@@ -0,0 +1,233 @@
+-}
+type Space :: *
+data Space = Space {
+  {-|
+  Skips zero or more (insignificant) whitespace characters as well as comments.
+
+  The implementation of this parser depends on whether 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true: 
+  when it is, this parser may change based on the use of the alter combinator. 
+
+  This parser will always use the hide combinator as to not appear as a valid alternative in an error message: 
+  it's likely always the case whitespace can be added at any given time, but that doesn't make it a useful suggestion unless it is significant.
+  -}
+    whiteSpace :: !(Parsec ())
+  {-|
+  Skips zero or more comments.
+
+  The implementation of this combinator does not vary with 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent'. 
+  It will use the hide combinator as to not appear as a valid alternative in an error message: 
+  adding a comment is often legal, 
+  but not a useful solution for how to make the input syntactically valid.
+  -}
+  , skipComments :: !(Parsec ())
+  {-|
+  This combinator changes how lexemes parse whitespace for the duration of a given parser.
+
+  So long as 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true, 
+  this combinator will be able to locally change the definition of whitespace during the given parser.
+
+  === __Examples__
+  * In indentation sensitive languages, the indentation sensitivity is often ignored within parentheses or braces. 
+  In these cases, 
+  @parens (alter withNewLine p)@ 
+  would allow unrestricted newlines within parentheses.
+  -}
+  , alter :: forall a. Desc.CharPredicate -> Parsec a -> Parsec a
+  {-|
+  This parser initialises the whitespace used by the lexer when 
+  'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true.
+
+  The whitespace is set to the implementation given by the lexical description.
+  This parser must be used, by fully or otherwise, 
+  as the first thing the global parser does or an UnfilledRegisterException will occur.
+
+  See 'alter' for how to change whitespace during a parse.
+  -}
+  , initSpace :: Parsec ()
+  }
+
+---------------------------------------------------------------------------------------------------
+-- Whitespace parsers setup
+
+{-|
+A global ref to the mapping from threads to their corresponding whitespace parsers.
+
+Implemented in terms of `unsafePerformIO` -- use with caution!
+-}
+-- TODO: use a HashMap instead of a Map?
+{-# NOINLINE wsImplMap #-}
+wsImplMap :: IORef (Map ThreadId (Parsec ()))
+wsImplMap = unsafePerformIO (newIORef Map.empty)
+
+
+{-| 
+Get the whitespace parser for this thread.
+
+This parser consumes no input and always succeeds.
+Throws a ghc `error` if the whitespace parser has not been initialised in `wsImplMap`
+-}
+getWs :: Parsec (Parsec ())
+getWs = Parsec $ \st good _ ->
+  do  ws <- unsafeIOToRT (Map.lookup <$> myThreadId <*> readIORef wsImplMap)
```

Comment:
This is going to be super expensive a lot of the time... Ideally we can just collect our Ref from the map and cache it for thread-local reuse. But now that I think about it, I'm not entirely sure how...

Reply:
Ah right. The expensive part, I think, is the atomic modify of the map. Getting from the map is probably still quite cheap. So, what I'm wondering is whether or not it's faster to store mutable references in the map, so that we aren't altering the map structure whenever we need to update our space, at the cost of reading two references...

Reply:
I guess an atomic modify is still needed on initialisation (so the thread can register itself with the map), but from then on can assume no other thread is touching its reference in the map?

Reply:
exactly